### ISSUE TITLE
Document that tests should create test data, rather than storing as files in the python path.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6535]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6535|
|Type|enhancement|
|Created|2013-05-31 20:50:25Z|
|Branch|https://github.com/twisted/twisted/tree/docs-test-data-6535|

```
<exarkun> keeping data in the source tree should be avoided where possible
<exarkun> I don't think anyone knows how to avoid it in _all_ cases, but where it's obvious how to do so it seems worthwhile.
<exarkun> And dumping data into a temporary path during the test run opens up more testing opportunities
<exarkun> If the data is part of the shipped source, then it's hard to rely on some of the less obvious properties (permissions, for example)
<exarkun> There's a handful of tests that fail when run against an installed copy of Twisted.  I dunno if this test is one of them, but it suffers from the same problem as the failing tests do, even if it doesn't have the same symptom
```

<details><summary>Searchable metadata</summary>

```
trac-id__6535 6535
type__enhancement enhancement
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_docs_test_data_6535 branches/docs-test-data-6535
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1370033425000000 1370033425000000
changetime__1447331250438646 1447331250438646
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_1"></a>|@Julian commented|
|-|-|

As per some discussion on [#5787](https://github.com/twisted/twisted/issues/5787) (and some more on IRC with exarkun for which this is a rough summary), we should consider test cases for trial to be an exception on a case-by-case basis when a separate module (to discover) might be necessary. (Though of course defining the auxiliary test code on the test case that needs it is preferable when possible anyhow).

[the rationale being that it is beneficial to be able to invoke `bin/trial` on the test modules besides their use in the relevant test]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @adiroiban|
|-|-|

I will try to solve this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban removed owner|
|-|-|

I have pushed an initial change to the documentation to talk about test data.

Please check and suggest changes.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this.

Few minor nitpicks:

- Line 238 needs to be brought onto the previous line (to match 1 sentence per line). Lines 242 and 243 also need to be modified to do this.

- There's two instances where "getThing" is typo'd as "getThink" in the code example (lines 269 and 274)

- I think lines 280 and so need to be reformed into a single paragraph or maybe two -- right now each sentence is its own paragraph.

- "Test" -> "Tests" at line 291

- I don't like the example docstring at line 303. It should mention keys.Key or so, rather than "it".

Please fix these up and merge.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Thanks for the review. Will fix and merge. Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46155]) Merge docs-test-data-6535: Update dev documentation to talk about test data and have example with task.Clock.

Author: adiroiban
Reviewer: hawkowl
Fixes: [#6535](https://github.com/twisted/twisted/issues/6535)
